File: go-binary/projects.go
package main

import (
	"context"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"net/url"
)

const resourceManagerSearchURL = "https://cloudresourcemanager.googleapis.com/v3/projects:search"

type rmProject struct {
	Name        string `json:"name"`
	ProjectID   string `json:"projectId"`
	DisplayName string `json:"displayName"`
	State       string `json:"state"`
}

type rmSearchResponse struct {
	Projects      []rmProject `json:"projects"`
	NextPageToken string      `json:"nextPageToken"`
}

// listOrgProjects returns the project_ids of every ACTIVE project under the
// given organization. Hand-rolled REST so the X-Goog-User-Project header
// (already injected by the http.Client transport) governs quota for these
// calls too.
func listOrgProjects(ctx context.Context, client *http.Client, orgID string) ([]string, error) {
	query := fmt.Sprintf("parent:organizations/%s state:ACTIVE", orgID)

	var ids []string
	pageToken := ""

	for {
		params := url.Values{}
		params.Set("query", query)
		if pageToken != "" {
			params.Set("pageToken", pageToken)
		}

		req, err := http.NewRequestWithContext(ctx, http.MethodGet,
			resourceManagerSearchURL+"?"+params.Encode(), nil)
		if err != nil {
			return nil, err
		}

		resp, err := client.Do(req)
		if err != nil {
			return nil, fmt.Errorf("resource manager request failed: %w", err)
		}
		body, err := io.ReadAll(resp.Body)
		resp.Body.Close()
		if err != nil {
			return nil, fmt.Errorf("reading resource manager response: %w", err)
		}

		if resp.StatusCode != http.StatusOK {
			return nil, fmt.Errorf("resource manager returned HTTP %d: %s",
				resp.StatusCode, truncate(string(body), 400))
		}

		var page rmSearchResponse
		if err := json.Unmarshal(body, &page); err != nil {
			return nil, fmt.Errorf("decoding resource manager response: %w", err)
		}

		for _, p := range page.Projects {
			if p.ProjectID == "" {
				continue
			}
			ids = append(ids, p.ProjectID)
		}

		if page.NextPageToken == "" {
			break
		}
		pageToken = page.NextPageToken
	}

	return ids, nil
}

func truncate(s string, n int) string {
	if len(s) <= n {
		return s
	}
	return s[:n] + "..."
}

File: python-gcloud/gcp_appliance_status/projects.py
"""Discover GCP projects belonging to an organization."""

import sys
import time

from google.cloud import resourcemanager_v3


def _log(msg: str) -> None:
    print(msg, file=sys.stderr, flush=True)


def list_org_projects(org_id: str) -> list[dict]:
    """List all active projects under an organization.

    Args:
        org_id: GCP organization ID (numeric string).

    Returns:
        List of dicts with project_id, name, and state.
    """
    _log(f"[projects] cloudresourcemanager.googleapis.com searchProjects "
         f"org={org_id}")
    t0 = time.monotonic()
    client = resourcemanager_v3.ProjectsClient()
    request = resourcemanager_v3.SearchProjectsRequest(
        query=f"parent:organizations/{org_id} state:ACTIVE"
    )

    projects = []
    for project in client.search_projects(request=request):
        projects.append({
            "project_id": project.project_id,
            "name": project.display_name,
            "state": project.state.name,
        })
        if len(projects) % 50 == 0:
            _log(f"[projects] streamed {len(projects)} project(s) so far...")

    _log(f"[projects] searchProjects returned {len(projects)} project(s) in "
         f"{time.monotonic() - t0:.2f}s")
    return projects
